teatest/teatest.go

```diff
@@ -0,0 +1,262 @@
+	tm := &TestModel{
+		in:      bytes.NewBuffer(nil),
+		out:     bytes.NewBuffer(nil),
+		modelCh: make(chan tea.Model, 1),
+		doneCh:  make(chan bool, 1),
+	}
+
+	tm.program = tea.NewProgram(
+		m,
+		tea.WithInput(tm.in),
+		tea.WithOutput(safe(tm.out)),
+		tea.WithoutSignals(),
+	)
+
+	interruptions := make(chan os.Signal, 1)
+	signal.Notify(interruptions, syscall.SIGINT)
+	go func() {
+		m, err := tm.program.Run()
+		if err != nil {
+			tb.Fatalf("app failed: %s", err)
+		}
+		tm.doneCh <- true
+		tm.modelCh <- m
+	}()
+	go func() {
+		<-interruptions
+		signal.Stop(interruptions)
+		tb.Log("interrupted")
+		tm.program.Quit()
+	}()
+
+	var opts TestModelOptions
+	for _, opt := range options {
+		opt(&opts)
+	}
+
+	if opts.size.Width != 0 {
+		tm.program.Send(opts.size)
+	}
+	return tm
+}
+
+func (tm *TestModel) waitDone() {
+	tm.done.Do(func() {
+		<-tm.doneCh
+	})
+}
+
+// FinalModel returns the resulting model, resulting from program.Run().
+// This method only returns once the program has finished running.
+func (tm *TestModel) FinalModel() tea.Model {
+	tm.waitDone()
+	select {
+	case m := <-tm.modelCh:
+		tm.model = m
+		return tm.model
+	default:
+		return tm.model
+	}
+}
+
+// Output returns the program's output io.Reader.
+func (tm *TestModel) Output() io.Reader {
+	return tm.out
+}
+
+// Send sends messages to the underlying program.
+func (tm *TestModel) Send(m tea.Msg) {
+	tm.program.Send(m)
+}
+
+// Quit quits the program and releases the terminal.
+func (tm *TestModel) Quit() error {
+	tm.program.Quit()
+	tm.program.Wait()
+	return tm.program.ReleaseTerminal()
+}
+
+// Type types the given text into the given program.
+func (tm *TestModel) Type(s string) {
+	for _, c := range []byte(s) {
+		tm.program.Send(tea.KeyMsg{
+			Runes: []rune{rune(c)},
+			Type:  tea.KeyRunes,
+		})
+	}
+}
+
+var update = flag.Bool("update", false, "update .golden files")
+
+// RequireEqualOutput is a helper function to assert the given output is
+// the expected from the golden files, printing its diff in case it is not.
+//
+// Important: this uses the system `diff` tool.
+//
+// You can update the golden files by running your tests with the -update flag.
+func RequireEqualOutput(tb testing.TB, out []byte) {
+	tb.Helper()
+
+	golden := filepath.Join("testdata", tb.Name()+".golden")
+	if *update {
+		if err := os.MkdirAll(filepath.Dir(golden), 0o755); err != nil { // nolint: gomnd
+			tb.Fatal(err)
+		}
+		if err := os.WriteFile(golden, out, 0o600); err != nil { // nolint: gomnd
+			tb.Fatal(err)
+		}
+	}
+
+	path := filepath.Join(tb.TempDir(), tb.Name()+".out")
+	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil { // nolint: gomnd
+		tb.Fatal(err)
+	}
+	if err := os.WriteFile(path, out, 0o600); err != nil { // nolint: gomnd
+		tb.Fatal(err)
+	}
+
+	// inspired by https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/cmd/internal/diff/diff.go;l=18
+	diff, err := exec.Command(
+		"diff",
+		"--minimal",
+		"--ignore-trailing-space",
```

Comment:
this is causing an error on mine when running `examples/simple`
```
=== RUN   TestApp
    main_test.go:41: output does not match, diff:
        
        diff: unrecognized option `--ignore-trailing-space'
        diff: Try `diff --help' for more information.
--- FAIL: TestApp (1.22s)
=== RUN   TestAppInteractive
--- PASS: TestAppInteractive (3.07s)
FAIL
FAIL	examples/simple	4.292s
FAIL
```

deleting this line fixed it, but not sure if it caused other side effects elsewhere yet.

Reply:
what's the output of `diff --version`?

Reply:
`diff (GNU diffutils) 2.8.1`

Reply:
hmmm mine is `diff (GNU diffutils) 3.8`... that explains it. Will look into it

Reply:
Kk let me know if there's anything I can help with

Reply:
To avoid relying on external diff tools, consider using an existing Go package to implement diffs. github.com/sergi/go-diff and github.com/pkg/diff are popular.